### ISSUE TITLE
useInject hook

### ISSUE THREAD

**Issue by luisherranz** [mobxjs/mobx-react-lite#6]
What do you think of a `useInject` hook?

Something like this (copied from the [`inject` example of `mobx-react`](https://github.com/mobxjs/mobx-react#customizing-inject)):
```js
const user = observable({
  name: "Noa"
});

const NameDisplayer = () => {
  const { name } = useInject(stores => ({
    name: stores.userStore.name
  }));
  return <h1>{name}</h1>;
};

const App = () => (
  <Provider userStore={user}>
    <NameDisplayer />
  </Provider>
);

ReactDOM.render(<App />, document.body);
```

We really like the `inject(mapper)` pattern and it's our default way to pass state to our components. It also avoids the need to wrap components with `observer`. 

It could use the next context API, and it could also include a `Provider` so people don't have to import the store/context time after time. And actually, I'd love it to be renamed to `useStores`. Much clearer:
```js
const NameDisplayer = () => {
  const { name } = useStores(stores => ({
    name: stores.userStore.name
  }));
  return <h1>{name}</h1>;
};
```

Any thoughts?

**Comment by FredyC** [comment]
I've just updated a documentation which explains why there is no Provider/inject. Please read that and let me know if you still think this is really needed :) In my opinion, it's not in the scope of the package anymore.

**Comment by luisherranz** [comment]
Following your example, using `useContext` instead of `useInject` would be:

```js
const userStore = observable({
  name: "Noa"
});

const StoresContext = React.createContext({ userStore });

const NameDisplayer = observer(() => {
  const { userStore } = useContext(StoresContext);
  const { name } = stores.userStore.name;
  return <h1>{name}</h1>;
});

const App = () => (
  <NameDisplayer />
);

ReactDOM.render(<App />, document.body);
```

Even if you don't want to include the `Provider`, we could improve the syntax with a `useInject` hook that provides the computation inside. Like this:

```js
const NameDisplayer = () => {
  const { name } = useInject(StoresContext, stores => ({
    name: stores.userStore.name,
  });
  return <h1>{name}</h1>;
};
```

We also use the function mapper a lot to prepare the props. This is a silly example but you get the point:
```js
const NameDisplayer = () => {
  const { name } = useInject(StoresContext, stores => ({
    hasName: !!stores.userStore.name,
    name: stores.userStore.name,
  });
  return hasName ? <h1>{name}</h1> : <h1>Sorry, no name!</h1>;
};
```

Of course all this (provider, hook) could be done in userland but it's great to have some standard ways to do things :)

**Comment by FredyC** [comment]
Why is it such a problem to do this? I mean there is no saving in a matter of LOC and on contrary, there is an extra overhead of a mapping function that would be executed on each component render.

```js
const { userStore } = useContext(StoresContext);
const { name } = userStore.name;
const hasName = !!name
```

Either way, something like `useInject` feels way too generic and I am sure it's only a matter of time before it appears in one of the [many packages](https://github.com/rehooks/awesome-react-hooks) that popped up.

Btw, there is some information that official release of hooks might include such mapping functionality natively to prevent rerendering when there is a change in a part of context you are not interested in.

**Comment by luisherranz** [comment]
I agree that a `useStores` hook could be replaced for these three operations:
- Using `observer()` in your component
- Using `import` to pass your StoresContext
- Using `useContext` to inject StoresContext

But I'm all in for patterns that make developer intentions easy to read, reduce the boilerplate and create standards among developers. It's only my opinion tho :)

> Either way, something like useInject feels way too generic and I am sure it's only a matter of time before it appears in one of the many packages that popped up.

Sure, if you think it's not general enough for this package it could go in a different `mobx-react-hooks` package. No problem with that.

**Comment by FredyC** [comment]
I am not sure I understand. First, the `observer` is not related to how you get the stores in the component. You might pass it through the props or have `useObservable` for a local state.

So ultimately if you import either `useContext` or `useInject` there is really no difference ❔ 

> Sure, if you think it's not general enough for this package it could go in a different `mobx-react-hooks` package. No problem with that.

But it's not related to MobX at all :) There is no observable _magic_ in a mapping output from the context ;)

**Comment by mweststrate** [comment]
> But it's not related to MobX at all :) There is no observable magic in a mapping output from the context ;)

Well, actually, there is, the mapping function it self is reactive; so that in case of `stores => ({    name: stores.userStore.name,  });` the stores are passed in as fresh props to the decorated component if `stores.userStore` or `userStore.name` property changes.

That being said, I think in most cases the selection can be done just inside the component (or smaller stores could be injected), I don't think there are much downsides to that, so I don't think we need `useInject`

**Comment by FredyC** [comment]
> That being said, I think in most cases the selection can be done just inside the component (or smaller stores could be injected), I don't think there are much downsides to that, so I don't think we need `useInject`

I agree. It will gain reactivity for "free" just by using properties from context directly. And for a more complex computation there will be [useComputed](https://github.com/mobxjs/mobx-react-lite/issues/4).

**Comment by luisherranz** [comment]
> Well, actually, there is, the mapping function it self is reactive; so that in case of stores => ({ name: stores.userStore.name, }); the stores are passed in as fresh props to the decorated component if stores.userStore or userStore.name property changes.

That's it, the mapper `stores => ({ name: stores.userStore.name, })` becomes the computation, so you don't need to wrap your component with `observer`.

That being said, we'll think about the syntax and if we still feel we prefer the old `inject` way to abstract `observer` and reduce the import boilerplate, we will create another package for that. 

Thanks guys!

**Comment by FredyC** [comment]
I can't help it, but I just find that from these two the latter one is so much easier to comprehend at a first glance and it will work the same way.

```tsx
const NameDisplayer = () => {
  const { name } = useInject(StoresContext, stores => ({
    hasName: !!stores.userStore.name,
    name: stores.userStore.name,
  });
  return hasName ? <h1>{name}</h1> : <h1>Sorry, no name!</h1>;
};
```

// vs

```jsx
const NameDisplayer = observer(() => {
  const { userStore } = useContext(StoresContext);
  const { name } = userStore.name;
  const hasName = !!name
  return hasName ? <h1>{name}</h1> : <h1>Sorry, no name!</h1>;
});
```

If you feel that extra import is a _boilerplate_ then I am just like 😮 Either way, have fun discovering what works best for you, that's what matters the most.

**Comment by FredyC** [comment]
@luisherranz Well, the `mobx-react` certainly won't go away in a near future and the `Provider/inject` will still work even with components using hooks. I don't think you need to jump the gun and switch to hooks just because of this hype. Although it's definitely more convenient than HOC or decorator.

But I understand your use case now and I am willing to accept PR for such hook. Please include tests and documentation in that case. And to kickstart it if you are not a friend with TypeScript, here goes the declaration that should work just fine.

```ts
export function useStore<T, M>(context: React.Context<T>, mapper: (store: T) => M): M {
    
}
```

**Comment by luisherranz** [comment]
> @luisherranz Well, the mobx-react certainly won't go away in a near future and the Provider/inject will still work even with components using hooks

Problem with the current `inject` is that the old context API would be removed in React 17.

> I don't think you need to jump the gun and switch to hooks just because of this hype. 

I know, but as `inject` will eventually die, we're looking for a long-term pattern.

Anyway, we know hooks aren't perfect. The thing I like about hooks is that they don't inject extra components in the React tree (`<inject-NameDisplayer>` in this case). And they think I dislike is that your components are not as dumb as they use to be!

```js
const NameDisplayer = ({ name, hasName }) => (
  hasName ? <h1>{name}</h1> : <h1>Sorry, no name!</h1>
);

export default inject(stores => ({
  hasName: !!stores.userStore.name,
  name: stores.userStore.name,
}))(NameDisplayer);
```

But I haven't found a way to overcome those two issues yet 😅 

> But I understand your use case now and I am willing to accept PR for such hook. 

Many thanks for the offer ☺️  

But this made me think more about the issue, so let me do it first in our app. As soon as the React team releases a non-alpha version of 16.7.0, we'll try some patterns here to see what feels just right. Then, we'll see if that pattern may be a good fit for the mobx community and if so, we'll do a PR with full docs & testing.

I'll reopen the issue when that happens.

**Comment by mweststrate** [comment]
I think we can switch to standard context implementation at some point,
generally I prefer to keep the legacy implementation as long as it is
available. Not because it is better, but because it avoids major version
number churn / people needing to find the right version. As long as we keep
the legacy implementation, mobx-react users with old versions of React
still receive all updates without needing backports :)

Op ma 12 nov. 2018 om 13:28 schreef Daniel K. <notifications@github.com>:

**Comment by luisherranz** [comment]
> You could even turn that into your own HOC, no restrains here really.

Sure, but then you're back to the "extra components in your tree" problem 😅 

> generally I prefer to keep the legacy implementation as long as it is
available

The old context API will be removed in React 17.0, which is great. Buuuut it can't be used in Strict-Mode (React Async/Concurrent), which should be available in React 16.7.0, which is bad:
`Warning: Legacy context API has been detected within a strict-mode tree`

So, although it may look like it will be around yet for a while, people who want to start using the new React features like Time Slicing or Suspense won't be able to use the current `mobx-react`.

But hey, that's what this package is for, right? :D
